### PR TITLE
Make abce more pythonic

### PR DESCRIPTION
**Pull request by DavoudTaghawiNejad**
- Make init function more pythonic
- Make Firm more pythonic and remove MultiproductFirms

### CODE REVIEW

File: abce/agent.py [hunk 1]
```diff
@@ -156,21 +156,36 @@ def __init__(self, id, agent_parameters, simulation_parameters, group, trade_log
...
+
+            def main():
+                sim = Simulation()
+                students = sim.build_agents(Student, 'student',
+                                            agent_parameters=[{'age': 12, lazy: True},
+                                                              {'age': 12, lazy: True},
+                                                              {'age': 13, lazy: False},
+                                                              {'age': 14, lazy: True}],
+                                            rounds=50,
+                                            school_size=990)
```
Comment (DavoudTaghawiNejad) [inline abce/agent.py]: The rationale for this API change is that in the previous version Agent.init(self, simulation_parameter, agent_parameter), even when no parameters where specified. Further, the modeler had to remember that init is called with two parameters no matter what and he would have to remember the order. 
When I presented abcEconomics in two weeks ago this was flagged as confusing. The new api, provides the parameters specified with the name they were specified in build_agents. Nothing to remember. 
The new api leads also to less boilerplate as we don't need to reference dictionaries in Agent.init anymore.
The new api makes it easy to read agents from an excel sheet or csv file::
 
    import csv  
with open('chess_set.csv') as csvfile:  
    reader = csv.DictReader(csvfile)
agents = sim.build_agents(Agent, 'agent', world_size=(8, 8), agent_parameter=reader)

Comment (rht) [reply abce/agent.py]: > and he would have to remember the order

Both `agent_parameter` and `parameter` are named arguments with default values, so ordering is not necessary, but helpful for consistent code.

File: abce/__init__.py [hunk 2]
```diff
@@ -391,32 +393,36 @@ def build_agents(self, AgentClass, group_name, number=None,
...
                                                 number=1,
-                                                parameters={'rounds':
-                                                             num_rounds})
+                                                rounds=num_rounds)
         """
         assert number is None or agent_parameters is None, \
             'either set number or agent_parameters in build_agents'
+        if agent_parameters is None:
+            agent_parameters = {}
+        if parameters is None:
```
Comment (rht) [inline abce/__init__.py]: This check isn't necessary, since there is no longer `parameters`.

Comment (rht) [reply abce/__init__.py]: Ah nvm, I looked at the init of the docstring section, instead of the actual def line.

File: abce/agent.py (hunk 1 above)
Comment (rht) [reply abce/agent.py]: > > and he would have to remember the order
> Both agent_parameter and parameter are named arguments with default values, so ordering is not necessary, but helpful for consistent code.

Ah this is not the case; right, the order is necessary.

File: abce/multiprocess.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def insert_or_append(self, group, ids, Agent, simulation_parameters, agent_param
                 agent = Agent(id, simulation_parameters, ap, **agent_arguments)
                 agent._send = agent._send_multiprocessing
                 agent._out = defaultdict(list)
-                agent.init(simulation_parameters, ap)
+                agent.init(**ChainMap(simulation_parameters, ap))
```
Comment (rht) [inline abce/multiprocess.py]: Why is the line not `agent.init(ap, **simulation_parameters)`?

Comment (rht) [reply abce/multiprocess.py]: From this line, my understanding is that even `agent_parameters` is no longer a required argument?

**Comment by rht** [comment]
It should be emphasized that the `Agent.init()` convention change here is not breaking any existing code.

**Comment by rht** [comment]
All the commits here are also present in #115, so I will just merge that one.

**Pull request closed without merging**